README.md
Crypto-Binary
=============
Assemble and disassemble binary messages used in cryptocoin applications.

MessageBuilder
--------------
Assemble a binary message out of chunks of data. All the methods return the builder itself, so are chainable. General usage is to use the various `put*` methods and then call `raw()` to get the output as a Buffer.

```js
var b = new MessageBuilder();
b.putInt8(255).putInt16(0xbeef);
b.put(new Buffer([1,2,3,4,5]));
console.log(b.raw().toString('hex')); // 'ffbeef0102030405'
```

* **putInt[8,16,32]**: Append an integer of the specified bit length to the message
* **put**: Append raw data to the message. If a number less than 255 is passed, acts like `putInt8`, otherwise input is expected to be a Buffer, which is appended to the message
* **putString**: Convert a string into ASCII-representation of the characters and append that to the message
* **putVarInt**: Convert a number into the Bitcoin [variable length integer](https://en.bitcoin.it/wiki/Protocol_specification#Variable_length_integer) notation and append it to the message
* **putVarString**: Convert a string into the Bitcoin [variable length string](https://en.bitcoin.it/wiki/Protocol_specification#Variable_length_string) notation and append it to the message


MessageParser
-------------
Given a string of binary data saved in a buffer, walk through it looking for structured data. Each of the methods returns the specified data from the current point in the data, and increments the pointer to the end of the data extracted. The point of this class is to allow you to not have to constantly check "is there enough data in the buffer to complete my next action?" or "did the last action fail?" and defer those checks to the end of parsing, making for cleaner code.

```js
var p = new MessageParser(rawData);
var itemNum = p.readVarInt();
for (var i = 0; i < itemNum; i++) {
  items.push(s.readUInt32LE());
}
if (p.hasFailed) {
  throw new Error('Ran out of data before this point!');
}
```

* **readUInt8**, **readUInt[16,32]LE**: Match the NodeJS Buffer methods of the same name; read an integer from the current pointer location of of the specified bit-size
* **readVarInt**: Grab a number encoded in the Bitcoin [variable length integer](https://en.bitcoin.it/wiki/Protocol_specification#Variable_length_integer) notation from the data stream
* **readVarString**: Grab a string encoded in the Bitcoin [variable length string](https://en.bitcoin.it/wiki/Protocol_specification#Variable_length_string) notation from the data stream
* **raw**: Grab a number of raw bytes from the data stream and return it as a Buffer.
* **incrPointer**, **setPointer**: Manipulate the current cursor location, either by incrementing it by a certain amount, or setting it to a specific absolute value

The MessageParser object has a `hasFailed` property, which is set to `true` if the actions request cannot be completed (pointer goes beyond the length of the buffer, usually). There is also a `failedStack` property that holds a stack trace from when the first failure occurred (in case you want to diagnose when in the process it failed)


CHANGELOG.md
0.1.1 / 2015-06-09
------------------
* Make buffer grow as needed, with a configurable maximum size (#1)

0.1.0 / 2014-02-03
------------------
* Initial release


test/build.test.js
var Builder = require('../lib/Message').MessageBuilder;
var assert = require("assert")

describe('Builder', function() {
  var b = false;
  beforeEach(function() {
    b = new Builder();
  });
  it('should properly build raw bytes', function() {
    b.put(1).put(2).put(3);
    assert.equal(b.raw().toString('hex'), '010203');
  });
  it('should properly build raw Buffers', function() {
    b.put(new Buffer([1])).put(new Buffer([2])).put(new Buffer([3]));
    assert.equal(b.raw().toString('hex'), '010203');
  });
  it('should properly pad raw bytes', function() {
    b.put(1).pad(5).put(1);
    assert.equal(b.raw().toString('hex'), '01000000000001');
  });
  it('should properly build 8-bit ints', function() {
    b.putInt8(1).putInt8(2).putInt8(3);
    assert.equal(b.raw().toString('hex'), '010203');
  });
  it('should properly build 16-bit ints', function() {
    b.putInt16(1).putInt16(2).putInt16(3);
    assert.equal(b.raw().toString('hex'), '010002000300');
  });
  it('should properly build 32-bit ints', function() {
    b.putInt32(1).putInt32(2).putInt32(3);
    assert.equal(b.raw().toString('hex'), '010000000200000003000000');
  });
  it('should properly build strings', function() {
    b.putString('Hello');
    assert.equal(b.raw().toString('utf8'), 'Hello');
  });
  it('should properly add varInts', function() {
    b.putVarInt(0x50).putVarInt(0x1234).putVarInt(0xdeadbeef).putVarInt(new Buffer([1,2,3,4,5,6,7,8]));
    assert.equal(b.raw().toString('hex'), '50fd3412feefbeaddeff0102030405060708');
  });
  it('should properly add varStrings', function() {
    b.putVarString('Hello');
    assert.equal(b.raw().toString('hex'), '0548656c6c6f');
  });
  it('should grow as needed', function() {
    var i
    for (i=0; i<10000; i++) b.put(1)
    assert.equal(b.cursor, 10000)
    for (i=0; i<5000; i++) b.put(new Buffer([2, 3]))
    assert.equal(b.cursor, 20000)
    b.putInt8(255)
    assert.equal(b.cursor, 20001)
    var str = b.raw().toString('hex')
    assert.equal(str.length, 40002)
    assert(/^(01){10000}(0203){5000}ff$/.test(str))
  })
  it('should limit the maximum size', function() {
    var b = new Builder(15000)
    for (var i=0; i<15000; i++) b.put(1)
    assert.throws(function(){ b.put(1) }, /Message size is limited/)
  })
});


lib/Message.js
var spareBytes = 512;

var MessageBuilder = exports.MessageBuilder = function MessageBuilder(maxBytes) {
  this.maxBytes = maxBytes || 20000000;
  this.buffer = new Buffer(Math.min(maxBytes, 10000));
  this.cursor = 0;
};
MessageBuilder.prototype.raw = function raw() {
  var out = new Buffer(this.cursor);
  this.buffer.copy(out, 0, 0, this.cursor);
  return out;
};
MessageBuilder.prototype.pad = function pad(num) {
  var data = new Buffer(num);
  data.fill(0);
  return this.put(data);
};
MessageBuilder.prototype.put = function put(data) {
  if (typeof data == 'number' && data <= 255) {
    this.ensureSize(1);
    this.buffer[this.cursor] = data;
    this.cursor += 1;
    return this;
  }
  
  this.ensureSize(data.length);
  data.copy(this.buffer, this.cursor);
  this.cursor += data.length;
  return this;
};
MessageBuilder.prototype.putInt8 = function putInt8(num) {
  if (Buffer.isBuffer(num)) {
    return this.put(num.slice(0, 1));
  }
  return this.put(num);
}
MessageBuilder.prototype.putInt16 = function putInt16(num) {
  if (Buffer.isBuffer(num)) {
    return this.put(num.slice(0, 2));
  }
  var data = new Buffer(2);
  data.writeUInt16LE(num, 0);
  return this.put(data);
};
MessageBuilder.prototype.putInt32 = function putInt32(num) {
  if (Buffer.isBuffer(num)) {
    return this.put(num.slice(0, 4));
  } else if (typeof num.getMonth === 'function') {
    return this.putInt32(num.getTime()/1000); // Pull timestamp from Date object
  }
  var data = new Buffer(4);
  data.writeUInt32LE(num, 0);
  return this.put(data);
};
MessageBuilder.prototype.putInt64 = function putInt64(num) {
  if (Buffer.isBuffer(num)) {
    return this.put(num.slice(0, 8));
  } else if (typeof num.getMonth === 'function') {
    return this.putInt64(num.getTime()/1000); // Pull timestamp from Date object
  }
  // Pad a 32-bit number to fit in a 64-bit space
  var data = new Buffer(8);
  data.fill(0);
  data.writeUInt32LE(num, 0);
  return this.put(data);
};
MessageBuilder.prototype.putString = function putString(str) {
  var data = new Buffer(str.length);
  for(var i = 0; i < str.length; i++) {
    data[i] = str.charCodeAt(i);
  }
  return this.put(data);
};
MessageBuilder.prototype.putVarInt = function putVarInt(num) {
  if (num < 0xfd) {
    return this.put(num);
  } else if (num <= 0xffff) {
    return this.put(0xfd).putInt16(num);
  } else if (num <= 0xffffffff) {
    return this.put(0xfe).putInt32(num);
  } else {
    return this.put(0xff).putInt64(num);
  }
};
MessageBuilder.prototype.putVarString = function putVarString(str) {
  return this.putVarInt(str.length).putString(str);
};
MessageBuilder.prototype.ensureSize = function ensureSize(additionalBytes) {
  var requiredBytes = this.cursor + additionalBytes;

  if (requiredBytes > this.maxBytes) {
    throw new Error('Message size is limited to ' + this.maxBytes + ' bytes');
  } else if (requiredBytes > this.buffer.length) {
    var oldBuffer = this.buffer;
    this.buffer = new Buffer(Math.min(this.maxBytes, requiredBytes + spareBytes));
    oldBuffer.copy(this.buffer);
  }
  return this;
}


var MessageParser = exports.MessageParser = function MessageParser(raw) {
  Object.defineProperty(this, 'buffer', {
    enumerable: true,
    value: new Buffer(raw.length)
  });
  raw.copy(this.buffer);
  
  this.pointer = 0;
  this.hasFailed = false;
  this.failedStack = false;
};

MessageParser.prototype.markFailed = function markFailed() {
  if (this.hasFailed) return false;
  this.hasFailed = true;
  this.failedStack = new Error().stack;
  return true;
};

MessageParser.prototype.pointerCheck = function pointerCheck(num) {
  num = +num || 0;
  if (this.buffer.length < this.pointer+num) {
    this.markFailed();
    return false;
  }
};

MessageParser.prototype.incrPointer = function incrPointer(amount) {
  if (this.hasFailed) return false;
  if (typeof amount !== 'number') {
    this.markFailed();
    return false;
  }
  this.pointer += amount;
  this.pointerCheck();
};

MessageParser.prototype.setPointer = function setPointer(amount) {
  if (this.hasFailed) return false;
  if (typeof amount !== 'number') {
    this.markFailed();
    return false;
  }
  this.pointer = amount;
  this.pointerCheck();
};

MessageParser.prototype.readInt8 = function readInt8() {
  if (this.hasFailed || this.pointerCheck() === false) return false;
  var out = this.buffer[this.pointer];
  this.incrPointer(1);
  return out;
};

MessageParser.prototype.readUInt16LE = function readUInt16LE() {
  if (this.hasFailed || this.pointerCheck(2) === false) return false;
  var out = this.buffer.readUInt16LE(this.pointer);
  this.incrPointer(2);
  return out
};

MessageParser.prototype.readUInt32LE = function readUInt32LE() {
  if (this.hasFailed || this.pointerCheck(4) === false) return false;
  var out = this.buffer.readUInt32LE(this.pointer);
  this.incrPointer(4);
  return out
};

MessageParser.prototype.readVarInt = function readVarInt() {
  if (this.hasFailed || this.pointerCheck() === false) return false;
  var flag = this.readInt8();
  if (flag < 0xfd) {
    return flag;
  } else if (flag == 0xfd) {
    return this.readUInt16LE();
  } else if (flag == 0xfe) {
    return this.readUInt32LE();
  } else {
    return this.raw(8);
  }
};

MessageParser.prototype.readVarString = function readVarString() {
  if (this.hasFailed || this.pointerCheck() === false) return false;
  var length = this.readVarInt();
  var str = [];
  for (var i = 0; i < length; i++) {
    str.push(String.fromCharCode(this.readInt8()));
  }
  return str.join('');
}

MessageParser.prototype.raw = function raw(length) {
  if (this.hasFailed || this.pointerCheck(length) === false) return false;
  if (typeof length !== 'number') {
    this.markFailed();
    return false;
  }
  var out = new Buffer(length);
  this.buffer.copy(out, 0, this.pointer, this.pointer+length);
  this.incrPointer(length);
  return out;
}
